components/sup/src/manager/service/mod.rs

```diff
@@ -404,29 +409,49 @@ impl Service {
         self.supervisor.state == ProcessState::Down
     }
 
-    /// Compares the current state of the service to the current state of the census ring and
-    /// re-renders all templatable content to disk.
+    /// Compares the current state of the service to the current state of the census ring and the
+    /// user-config, and re-renders all templatable content to disk.
     ///
-    /// Returns true if any modifications were made.
+    /// Returns `true` if any modifications were made.
     fn update_templates(&mut self, census_ring: &CensusRing) -> bool {
         let census_group = census_ring.census_group_for(&self.service_group).expect(
             "Service update failed; unable to find own service group",
         );
-        let cfg_updated = self.cfg.update(census_group);
-        if cfg_updated || census_ring.changed() {
+        let cfg_updated_from_rumors = self.cfg.update(census_group);
+        let cfg_changed = cfg_updated_from_rumors || self.user_config_updated;
+        let something_to_do = cfg_changed || census_ring.changed();
+
+        if something_to_do {
```

Comment:
This became a bit hard to follow. Would pseudo-rust below be clearer? (`cufr` = `cfg_updated_from_rumors`, `sucu` = `self.user_config_updated`, `crc` = `census_ring.changed()`)

```
if sucu {
    if let Err(e) = self.cfg.load_user(&self.pkg) {…}
}
let (cfg_changed, reload, reconfigure) = {
    let ctx = …;
    if cufr || crc {
        (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))
    } else if sucu {
        (true, true, self.compile_configuration(&ctx))
    } else {
        (false, self.needs_reload, self.needs_reconfiguration)
    }
}
self.needs_reload = reload;
self.needs_reconfiguration = reconfigure;
self.user_config_updated = false;
/* return */ cfg_changed
```

Reply:
The `something_to_do` variable would be gone.

Reply:
I find the 3 value tuple even less clear tbh...

Reply:
Not sure what this "what?" is about, but to be clear - I don't actually want to have variables like `cufr` or `sucu`. I wrote it like that to avoid typing in the comment.

Reply:
I think that even with more descriptive names, it's hard to parse logically. There's just too much happening on one line: returning a tuple, ORing two values, calling functions..

Not sure it's an improvement.

Reply:
You don't need to put everything in one line. You could very well have:
```
let confg_changed = config_updated_from_rumors || self.user_config_updated;
let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
let needs_reconfigure = self.compile_configuration(&ctx);
(confg_changed, needs_reload || sucu, needs_reconfigure)
```

In general, what I wanted to propose is to simplify the flow of the code, instead of using nested `if`s and `match`es that are sometimes executed and sometimes not.

Reply:
In the above implementation though, we're calling `self.compile_hooks` even if we don't need to, which will lead to "Hooks recompiled" being printed over and over (this function gets executed [on every tick](https://github.com/kinvolk/habitat/blob/0a1d210328174ae822c13213e2462f6f37f1e94c/components/sup/src/manager/service/mod.rs#L294)).

I'm definitely open to simplifying the code but obviously it has to have the same behaviour.

The problem is that some of these functions have side effects so we can't just call them willy-nilly.

Reply:
The `self.compile_hooks(&ctx)` should only be executed when either `cfg_updated_from_rumors` is true or if `census_ring.changed()` returns true. So if this happens everytime, then something is wrong.